apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -137,12 +157,24 @@ private static TraceHeaderPropagatingHandler CreateTraceHeaderPropagatingHandler
         }
 
         /// <summary>
-        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>
+        /// and a <see cref="ShouldTraceRequest"/>.
         /// </summary>
         internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
         {
-            var accessor = provider.GetService<IHttpContextAccessor>();
-            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            var accessor = provider.GetServiceCheckNotNull<IHttpContextAccessor>();
+            var shouldTraceRequest = provider.GetServiceCheckNotNull<ShouldTraceRequest>();
+            var traceIdFactory = provider.GetServiceCheckNotNull<TraceIdFactory>();
+
+            string header = accessor.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            var traceHeaderContext = TraceHeaderContext.FromHeader(header);
+            if (traceHeaderContext.ShouldTrace == null && shouldTraceRequest.ShouldTrace != null)
+            {
+                string traceId = traceHeaderContext.TraceId ?? traceIdFactory.NextId();
+                ulong spanId = traceHeaderContext.SpanId ?? 0;
+                bool? shouldTrace = shouldTraceRequest.ShouldTrace(accessor.HttpContext?.Request);
```

Comment:
I had thought about this initially but couldn't find a compelling reason why we would need the context and thought it may make things a little more confusing.

However due to the "magic of HttpContext" if users really want it they can get it from the Request as it has a HttpContext field.